### PR TITLE
Coding guidelines for external contributions

### PR DESCRIPTION
**Pull request by ashmeenkaur**
### Description
Added coding guidelines for external contributions.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: docs/dev_guide.md [hunk 1]
```diff
@@ -1,31 +1,149 @@
...
+contribution.**
+
+Rollout of any new feature should be done in 2 phases:
+
+**1. Experimental Phase:**
+
+* New features should initially be introduced behind an experimental flag. These
+  flags should be prefixed with `--experimental-` and marked as hidden in the
+  help output.
+* This signals to users that the feature is in development and may change or be
```
Comment (Tulsishah) [inline docs/dev_guide.md:16]: If the feature is going to be hidden, then using the `--experimental` flag makes sense? I feel we should choose one approach: either create a hidden flag or a public flag with the `--experimental` prefix.

File: docs/dev_guide.md [hunk 2]
```diff
@@ -1,31 +1,149 @@
...
-1.  Add unit tests in
-    [validate_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/validate_test.go)
-    and
-    [rationalize_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/rationalize_test.go)
-    and composite tests in
-    [config_validation_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/config_validation_test.go),
-    [config_rationalization_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/config_rationalization_test.go)
-    and
-    [root_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/root_test.go)
+1. Declare the new param in
```
Comment (Tulsishah) [inline docs/dev_guide.md:40]: I feel adding dummy PR as reference will be more helpful here.

File: docs/dev_guide.md [hunk 3]
```diff
@@ -1,31 +1,149 @@
...
+* After the feature development is complete, developers should open a GitHub
+  issue (
+  see [example](https://github.com/GoogleCloudPlatform/gcsfuse/issues/793)) to
+  track the feature's progress towards production readiness. This issue should
+  outline the motivation for the feature, its design, and any known limitations.
+
+**2. Production Rollout:**
+
+* Once the feature has been thoroughly vetted in test environments for at least
+  a week and approved by the GCSFuse team, it will be rolled out for general
```
Comment (charith87) [inline docs/dev_guide.md:27]: do we need the mention for at least a week. Or can we remove it.

File: docs/dev_guide.md (hunk 1 above)
Comment (charith87) [inline docs/dev_guide.md:16]: Can we add they should add tests as well in the experimental phase to understand the expected behaviour and use-cases

File: docs/dev_guide.md [hunk 4]
```diff
@@ -1,31 +1,149 @@
...
+  help output.
+* This signals to users that the feature is in development and may change or be
+  removed.
+* After the feature development is complete, developers should open a GitHub
+  issue (
+  see [example](https://github.com/GoogleCloudPlatform/gcsfuse/issues/793)) to
+  track the feature's progress towards production readiness. This issue should
+  outline the motivation for the feature, its design, and any known limitations.
+
+**2. Production Rollout:**
```
Comment (charith87) [inline docs/dev_guide.md:24]: So will the old PR be merged and this is in a follow-up PR.

File: docs/dev_guide.md [hunk 5]
```diff
@@ -1,31 +1,149 @@
...
-    [config_validation_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/config_validation_test.go),
-    [config_rationalization_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/config_rationalization_test.go)
-    and
-    [root_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/root_test.go)
+1. Declare the new param in
+   [params.yaml](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/params.yaml#L4).
+   Refer to the documentation at the top of the file for guidance.
+2. Run `make build` from the project root to generate the required code.
+3. Add validations on the param value in
+   [validate.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/validate.go)
```
Comment (Tulsishah) [inline docs/dev_guide.md:45]: I feel remove the reference of master as it can be change with file change in master.

File: docs/dev_guide.md (hunk 1 above)
Comment (charith87) [reply docs/dev_guide.md:16]: I think it feels fine to have the prefix experimental as well. We can take the decision when to remove it based on whether it can be documented in public docs or not.

File: docs/dev_guide.md [hunk 6]
```diff
@@ -1,31 +1,149 @@
...
+Rollout of any new feature should be done in 2 phases:
+
+**1. Experimental Phase:**
+
+* New features should initially be introduced behind an experimental flag. These
+  flags should be prefixed with `--experimental-` and marked as hidden in the
+  help output.
+* This signals to users that the feature is in development and may change or be
+  removed.
+* After the feature development is complete, developers should open a GitHub
```
Comment (charith87) [inline docs/dev_guide.md:18]: Also should add some rationale on when to add composite or end to end tests.

File: docs/dev_guide.md (hunk 4 above)
Comment (ashmeenkaur) [reply docs/dev_guide.md:24]: Yes, do you think this needs to be explicitly mentioned?

File: docs/dev_guide.md (hunk 2 above)
Comment (ashmeenkaur) [reply docs/dev_guide.md:40]: I think adding a reference PR will confuse the customer as it is mostly auto-generated code. Let's not add it. Moreover, this change is not part of the PR and is showing up as a change due to formatting change (change in list numbering).

File: docs/dev_guide.md (hunk 5 above)
Comment (ashmeenkaur) [reply docs/dev_guide.md:45]: We are referring to specific files on master here. Perhaps if the files move, we need to refresh the guide too?

I am not inclined towards making this change in this PR  since this change is not part of the PR and is showing up as a change due to formatting change (change in list numbering).

File: docs/dev_guide.md [hunk 7]
```diff
@@ -1,31 +1,166 @@
...
-    [rationalize_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/rationalize_test.go)
-    and composite tests in
-    [config_validation_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/config_validation_test.go),
-    [config_rationalization_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/config_rationalization_test.go)
-    and
-    [root_test.go](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/root_test.go)
+1. Declare the new param in
+   [params.yaml](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/params.yaml#L4).
+   Refer to the documentation at the top of the file for guidance.
+2. Run `make build` from the project root to generate the required code.
```
Comment (kislaykishore) [inline docs/dev_guide.md:60]: Why are we explicitly assigning indices instead of letting the renderer compute them itself. If we let it auto-compute by assigning index 1. for every item, we don't have to renumber while adding/removing steps.

Comment (ashmeenkaur) [reply docs/dev_guide.md:60]: It was showing up as a warning on my IDE so I applied the suggested fix. I don't think this is an added toil as it is auto fixed by IDE. Let me know if you think this should be reverted.

**Review by Tulsishah — APPROVED** [review]

File: docs/dev_guide.md (hunk 4 above)
Comment (charith87) [reply docs/dev_guide.md:41]: Yea feel its easy to get confused what is production rollout for us vs them.
For the customers production rollout could mean its just available in a production version or next version.
For us, i guess the meaning here is for all users availability. ( this could be outside the next immediate version ).

File: docs/dev_guide.md [hunk 8]
```diff
@@ -1,31 +1,166 @@
...
+       ```
+    - To run a particular test add `-run` flag to the command:
+       ```shell
+       export TEST_NAME=<Enter particular test name you want to run>
+       export TEST_PACKAGE_NAME=<Enter package name here. For eg: operations>
+       export TEST_BUCKET_NAME=<Enter your bucket name here.>
+       GODEBUG=asyncpreemptoff=1 CGO_ENABLED=0 go test ./tools/integration_tests/$TEST_PACKAGE_NAME/... -p 1 -short --integrationTest -v --testbucket=$TEST_BUCKET_NAME --timeout=60m -run $TEST_NAME
+       ```
+4. **Run all tests as pre-submit:** You can run existing GCSFuse end-to-end
+   tests as a pre-submit by adding `execute-integration-tests` label to the pull
```
Comment (Tulsishah) [inline docs/dev_guide.md:161]: External user will not be able to add any label on the PR. We have to provide `go test` command or steps to run our e2e test script locally.

File: docs/dev_guide.md [hunk 9]
```diff
@@ -1,31 +1,166 @@
 # Dev Guide
 
-## Adding a new param in GCSFuse
+## Adding new features to GCSFuse
```
Comment (marcoa6) [inline docs/dev_guide.md:3]: Change to "Contributing to GCSfuse"

File: docs/dev_guide.md [hunk 10]
```diff
@@ -1,31 +1,166 @@
...
+
+**If you're interested in adding a new feature, please open a GitHub issue to
```
Comment (marcoa6) [inline docs/dev_guide.md:5]: If you are interested in contributing a new feature, enhancement, or bug fix, please open

File: docs/dev_guide.md [hunk 11]
```diff
@@ -1,31 +1,166 @@
...
+discuss your proposal with the GCSFuse team first before making the
+contribution.**
```
Comment (marcoa6) [inline docs/dev_guide.md:7]: 'Make sure to describe the use case (what are you trying to solve and why?), and the proposed solution.

File: docs/dev_guide.md [hunk 12]
```diff
@@ -1,31 +1,166 @@
...
+      the [testing guide](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/dev_guide.md#testing-of-new-features)
+      below.
+
+* After the feature development is complete, developers should open a GitHub
+  issue (
+  see [example](https://github.com/GoogleCloudPlatform/gcsfuse/issues/793)) to
+  track the feature's progress towards production readiness. This issue should
+  outline the motivation for the feature, its design, and any known limitations.
+
+**2. Production Rollout:**
```
Comment (marcoa6) [inline docs/dev_guide.md:41]: Change Production roll out to "GA Rollout"

File: docs/dev_guide.md (hunk 11 above)
Comment (marcoa6) [reply docs/dev_guide.md:7]: Also add: We make no guarantees that contribution requests are accepted, or will be rolled out as a Generally Available feature.

File: docs/dev_guide.md [hunk 13]
```diff
@@ -1,31 +1,166 @@
...
+
+* Once the feature has been thoroughly vetted in test environments and approved
+  by the GCSFuse team, it will be rolled out for general availability. The
```
Comment (marcoa6) [inline docs/dev_guide.md:44]: Once the feature has been thoroughly vetted in test environments and approved
  by the GCSFuse team, it will be rolled out for general availability. There is no guarantee that experimental features will become generally available.

File: docs/dev_guide.md [hunk 14]
```diff
@@ -1,31 +1,166 @@
...
-## Adding a new param in GCSFuse
+## Adding new features to GCSFuse
+
+**If you're interested in adding a new feature, please open a GitHub issue to
+discuss your proposal with the GCSFuse team first before making the
+contribution.**
+
+Rollout of any new feature should be done in 2 phases:
+
+**1. Experimental Phase:**
```
Comment (marcoa6) [inline docs/dev_guide.md:11]: Where will we keep/document features in experimental phase and their flags? I dont want to document them on public docs, so can we have a doc under contributing that shows all contributed features and their status? And ask users to update that as well?

File: docs/dev_guide.md (hunk 8 above)
Comment (ashmeenkaur) [reply docs/dev_guide.md:161]: Thanks for catching this. I will change to "Ask GCSFuse team to apply `execute-integration-tests` label.

File: docs/dev_guide.md (hunk 14 above)
Comment (ashmeenkaur) [reply docs/dev_guide.md:11]: IMO, this yaml file (which we use to generate flag/config code) should serve the purpose of documentation too: https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cfg/params.yaml

"experimental" prefix in the `flag-name:` and `hide-flag:` parameter should convey the status.

File: docs/dev_guide.md [hunk 15]
```diff
@@ -1,31 +1,183 @@
...
+you trying to solve and why?), and the proposed solution in the GitHub issue.
+
+**Note:** We make no guarantees that contribution requests are accepted, or will
+be rolled out as a Generally Available (GA) feature
+
+Rollout of any new feature should be done in 2 phases:
+
+**1. Experimental Phase:**
+
+* New features should initially be introduced behind an experimental flag. These
```
Comment (charith87) [inline docs/dev_guide.md:17]: Can we add make some text here to link to the section below of Adding a CLI/Config flag.

File: docs/dev_guide.md [hunk 16]
```diff
@@ -1,31 +1,183 @@
...
+  flags should be prefixed with `--experimental-` and marked as hidden in the
+  help output.
+* This signals to users that the feature is in development and may change or be
+  removed.
+* **Testing Requirements**:
+    * Thorough testing is a crucial requirement for any new features added to
+      GCSFuse even during the experimental phase. This helps ensure the
+      feature's stability, identify potential issues early, and prevent
+      regressions.
+    * Add unit tests for all new code written for the experimental feature. This
```
Comment (charith87) [inline docs/dev_guide.md:27]: Add link to the section below of unit tests

File: docs/dev_guide.md [hunk 17]
```diff
@@ -1,31 +1,183 @@
...
+      includes testing edge cases, boundary conditions, and error handling.
+    * Add composite tests to validate GCSFuse's end user functionality while
```
Comment (charith87) [inline docs/dev_guide.md:29]: Similar to composite and end to end.

File: docs/dev_guide.md [hunk 18]
```diff
@@ -1,31 +1,183 @@
...
+      below.
+      After the feature development is complete and the code is merged to the
+      mainline, the feature will roll out as an experimental feature in the next
+      GCSFuse release with the feature disabled by default.
+
+* After the feature development is complete and the code is merged to the
+  mainline, the feature will roll out as an experimental feature in the next
+  GCSFuse release with the feature disabled by default.
+
+* Developers should open a GitHub issue (
```
Comment (charith87) [inline docs/dev_guide.md:46]: Should the developers create a new github issue in addition to the one they started created before they started the experimental changes?

File: docs/dev_guide.md (hunk 8 above)
Comment (Tulsishah) [reply docs/dev_guide.md]: Although they'll not be able to see which test fails on their PR. They won't be able to access fusion link. Better to share command to tests on them self

Comment (ashmeenkaur) [reply docs/dev_guide.md]: Discussed offline, we won't be sharing the script with the customer due to following reasons:
1.  the tests are huge and there is cost associated with running the tests, expecting customers to run the whole test suite is not right.  
2. Some of our tests rely on the secrets stored in the project's secret manager which will fail for the customer.

So we are making it code reviewer responsibility to share the failure logs with the customer.

**Review by Tulsishah — DISMISSED** [review]

File: docs/dev_guide.md (hunk 18 above)
Comment (ashmeenkaur) [reply docs/dev_guide.md:46]: Yes proposing it to be a new github issue. Original issue for proposal can be closed after experimental implementation is complete.

**Review by Tulsishah — APPROVED** [review]

**Review by charith87 — APPROVED** [review]
Approving as overall seems fine from my side.
Incremental changes can be done as separate PRs as necessary.
